engine/resources/docker.go

```diff
@@ -0,0 +1,415 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/engine/traits"
+	"github.com/purpleidea/mgmt/util"
+)
+
+const (
+	// dockerClientVersion is the api version used by the client. See:
+	// https://docs.docker.com/develop/sdk/#api-version-matrix for more info.
+	dockerClientVersion = "v1.37"
+	// dockerContainerRunning is the running container state.
+	containerRunning = "running"
+	// dockerContainerStopped is the stopped container state.
+	containerStopped = "stopped"
+	// dockerContainerRemoved is the removed container state.
+	containerRemoved = "removed"
+	// ctxTimeout is the length of time, in seconds, before requests are
+	// cancelled in Validate.
+	ctxTimeout = 20
+)
+
+func init() {
+	engine.RegisterResource("docker", func() engine.Res { return &DockerRes{} })
+}
+
+// DockerRes is a docker account resource.
+type DockerRes struct {
+	traits.Base // add the base methods without re-implementation
+	traits.Edgeable
+
+	State string                     `yaml:"state"` // running, stopped, or removed
+	Image string                     `yaml:"image"` // docker image or image:tag
+	Cmd   []string                   `yaml:"cmd"`   // command(s) to run on the container
+	Env   []string                   `yaml:"env"`   // environment variables ("VAR=val")
+	Ports map[string]map[int64]int64 `yaml:"ports"` // e.g. {"tcp" => {80 => 8080},},
+
+	client *client.Client // docker api client
+
+	init *engine.Init
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *DockerRes) Default() engine.Res {
+	return &DockerRes{}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *DockerRes) Validate() error {
+	ctx, cancel := context.WithTimeout(context.Background(), ctxTimeout*time.Second)
+	defer cancel()
+
+	// validate state
+	if obj.State != containerRunning && obj.State != containerStopped && obj.State != containerRemoved {
+		return fmt.Errorf("state must be running stopped or removed")
+	}
+
+	// validate image
+	c, err := client.NewClient(client.DefaultDockerHost, dockerClientVersion, nil, nil)
+	if err != nil {
+		return fmt.Errorf("error creating docker client: %s", err)
+	}
+	resp, err := c.ImageSearch(ctx, obj.Image, types.ImageSearchOptions{Limit: 1})
+	if err != nil {
+		return fmt.Errorf("error searching for image: %s", err)
+	}
+	if len(resp) == 0 {
+		return fmt.Errorf("image: %s not found", obj.Image)
+	}
+	c.Close()
```

Comment:
If we care about the return value (probably an err) then use it, otherwise this should be a defer in the right place I think.